packages/wallet/src/TxBuilder/buildTx.ts

```diff
@@ -170,19 +173,70 @@ export class ObservableWalletTxBuilder implements TxBuilder {
   }
 
   async #validateOutputs(): Promise<TxOutValidationError[]> {
-    const outputValidations =
-      this.partialTxBody.outputs && (await this.#util.validateOutputs(this.partialTxBody.outputs));
-
-    const errors = [...(outputValidations?.entries() || [])]
-      .map(([txOut, validation]) => toOutputValidationError(txOut, validation))
-      .filter((err): err is OutputValidationTokenBundleSizeError | OutputValidationMinimumCoinError => !!err);
+    const errors = (
+      await Promise.all(this.partialTxBody.outputs?.map((output) => this.buildOutput(output).build()) || [])
+    ).flatMap((output) => (output.isValid ? [] : output.errors));
 
     if (errors.length > 0) {
       throw errors;
     }
     return [];
   }
 
+  async #addDelegationCertificates(): Promise<void> {
+    if (!this.#delegateConfig) {
+      // Delegation was not configured by user
+      return Promise.resolve();
+    }
+
+    const rewardAccounts = await firstValueFrom(this.#observableWallet.delegation.rewardAccounts$);
+
+    if (!rewardAccounts?.length) {
+      // This shouldn't happen
+      throw new IncompatibleWalletError();
+    }
+
+    // Discard previous delegation and prepare for new one
+    this.partialTxBody = { ...this.partialTxBody, certificates: [] };
+
+    for (const rewardAccount of rewardAccounts) {
+      const stakeKeyHash = Cardano.Ed25519KeyHash.fromRewardAccount(rewardAccount.address);
+      if (this.#delegateConfig.type === 'deregister') {
+        // Deregister scenario
+        if (rewardAccount.keyStatus === StakeKeyStatus.Unregistered) {
+          this.#logger.warn(
+            'Skipping stake key deregister. Reward account not registered',
+            rewardAccount.address,
+            rewardAccount.keyStatus
+          );
+        } else {
+          this.partialTxBody.certificates!.push({
+            __typename: Cardano.CertificateType.StakeKeyDeregistration,
+            stakeKeyHash
+          });
+        }
+      } else if (this.#delegateConfig.type === 'delegate') {
+        // Register and delegate scenario
+        if (rewardAccount.keyStatus !== StakeKeyStatus.Unregistered) {
+          this.#logger.warn(
```

Comment:
De-register deserves a warning because it's user's action that doesn't make sense. In this case re-delegating is a normal flow.

```suggestion
          this.#logger.debug(
```

Reply:
[ccdd790](https://github.com/input-output-hk/cardano-js-sdk/pull/427/commits/ccdd7900f9104273db3cfc67f1ca133dd7f6d9de)